### PR TITLE
refactor: Mark functions async in `aws/customResources` and `aws/deploy`

### PR DESCRIPTION
**Pull request by ifitzsimmons**
Refactored Promise-returning functions for:
* `lib/plugin/aws/customResources`
* `lib/plugin/aws/deploy`

I had to refactor a few tests after tagging functions with `async` keyword. I didn't change the functionality of the test, just the syntax.

Addresses: #8368

### CODE REVIEW

File: lib/plugins/aws/lib/naming.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ module.exports = {
       this.provider.serverless.service.provider.stackName &&
       typeof this.provider.serverless.service.provider.stackName === 'string'
     ) {
-      return `${this.provider.serverless.service.provider.stackName}`;
+      return this.provider.serverless.service.provider.stackName;
```
Comment (pgrzesik) [inline lib/plugins/aws/lib/naming.js:56]: good catch :+1:

File: test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js [hunk 2]
```diff
@@ -90,27 +90,25 @@ describe('extendedValidate', () => {
...
+    it('should not throw error if service has no functions and no service package', async () => {
       stateFileMock.service.functions = {};
       fileExistsSyncStub.returns(true);
       readFileSyncStub.returns(stateFileMock);
 
-      return awsDeploy.extendedValidate().then(() => {
-        expect(fileExistsSyncStub.calledOnce).to.equal(true);
-        expect(readFileSyncStub.calledOnce).to.equal(true);
-      });
+      await awsDeploy.extendedValidate();
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:98]: Does it have to be `await`-ed here? According to the changes to `extendedValidate`, that function is no longer async

File: test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js [hunk 3]
```diff
@@ -90,27 +90,25 @@ describe('extendedValidate', () => {
...
       stateFileMock.service.functions = {};
       awsDeploy.serverless.service.package.individually = true;
       fileExistsSyncStub.returns(true);
       readFileSyncStub.returns(stateFileMock);
 
-      return awsDeploy.extendedValidate().then(() => {
-        expect(fileExistsSyncStub.calledOnce).to.equal(true);
-        expect(readFileSyncStub.calledOnce).to.equal(true);
-      });
+      await awsDeploy.extendedValidate();
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:109]: Same as above

File: test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js [hunk 4]
```diff
@@ -156,18 +153,18 @@ describe('extendedValidate', () => {
...
       // const fileExistsSyncStub = sinon.stub(awsDeploy.serverless.utils, 'fileExistsSync');
       fileExistsSyncStub.onCall(0).returns(true);
       fileExistsSyncStub.onCall(1).returns(true);
       readFileSyncStub.returns(stateFileMock);
       awsDeploy.serverless.service.package.artifact = 'some/file.zip';
-      return awsDeploy.extendedValidate().then(() => {
-        delete awsDeploy.serverless.service.package.artifact;
-      });
+
+      await awsDeploy.extendedValidate();
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:163]: Same here

File: test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js [hunk 5]
```diff
@@ -185,13 +182,13 @@ describe('extendedValidate', () => {
...
       readFileSyncStub.returns(stateFileMock);
 
-      return awsDeploy.extendedValidate().then(() => {
-        const msg = [
-          "WARNING: Function first has timeout of 31 seconds, however, it's ",
-          "attached to API Gateway so it's automatically limited to 30 seconds.",
-        ].join('');
-        expect(awsDeploy.serverless.cli.log.firstCall.calledWithExactly(msg)).to.be.equal(true);
-      });
+      await awsDeploy.extendedValidate();
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:185]: same here

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @ifitzsimmons :bow: Overall it looks good, I've had a few minor comments, but nothing major :+1:

File: test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js (hunk 2 above)
Comment (ifitzsimmons) [reply test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:98]: Ah, good catch. I'll fix all of those `await`'s.

**Comment by ifitzsimmons** [comment]
@pgrzesik, I'm having a little trouble merging master back in to my branch. It looks like once I merge it, there are some 
strange errors that don't have much to do with my branch. At the very last, there weren't any manual merges required for any of those files.

image

**Comment by pgrzesik** [comment]
Hey @ifitzsimmons, sorry for the trouble. Yesterday we had some upgrades to internal dependencies which are most likely causing issues for you - reinstalling all deps locally after you merge master back or rebase your branch on top of it should do the trick. Also, we switched to `v2` of `prettier`, so there might be some minor conflicts for you to resolve.

**Comment by ifitzsimmons** [comment]
Okay cool, I wasn't sure how you felt about a rebase during the review process but I'll do that today.

**Comment by ifitzsimmons** [comment]
@pgrzesik I got dinged for a dip in code coverage for the following changes... can we override that github action?

image

image

I ran coverage locally for these files and nothing has changed between `master` and this branch as a sanity check.

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js [hunk 6]
```diff
@@ -126,11 +126,11 @@ describe('createStack', () => {
     it('should throw error if invalid stack name', () => {
       sandbox.stub(awsDeploy, 'create').resolves();
       sandbox.stub(awsDeploy.provider, 'request').resolves();
-      awsDeploy.serverless.service.service = 'service-name'.repeat(100);
+      awsDeploy.serverless.service.provider.stackName = 'service-name'.repeat(100);
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:129]: Is there a reason to change this test setup here?

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @ifitzsimmons for the changes. I've inspected the coverage changes and it's possible that I gave you the wrong suggestion with dropping the `async` from `extendedValidate`. It is used in a few places and possibly it's expected from that function to return a Promise (it's used in `AwsInvoke` where we see the dropped coverage). Would you be able to change it back and validate again if we're experiencing coverage drop? Sorry for the trouble :bow:

File: test/unit/lib/plugins/aws/deploy/lib/createStack.test.js (hunk 6 above)
Comment (ifitzsimmons) [reply test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:129]: Huh, I'm not sure how or why that snuck in there.... I did perform a rebase, but to answer your question - no, there's no reason that that should be there.

**Comment by ifitzsimmons** [comment]
> Thank you @ifitzsimmons for the changes. I've inspected the coverage changes and it's possible that I gave you the wrong suggestion with dropping the `async` from `extendedValidate`. It is used in a few places and possibly it's expected from that function to return a Promise (it's used in `AwsInvoke` where we see the dropped coverage). Would you be able to change it back and validate again if we're experiencing coverage drop? Sorry for the trouble 🙇

I think that maybe I caused that confusion. I removed all async/await logic associated with `extendedValidate()`. Before these changes, it returned `BBPromise.resolve()` without ever calling any functions that returned promises. I think your initial recommendation to remove async from extended validate was spot on.   Still having codecov issues since I removed the call to BBPromise.resolve. Adding a test case now to increase codecov

**Comment by ifitzsimmons** [comment]
@pgrzesik Let me know if you want me to squash all of these commits before merging.

**Comment by ifitzsimmons** [comment]
@pgrzesik Am I missing something here. When I click on the codecov report, I am seeing that extendedValidate() is the violator. But when I run codecoverage locally, my branch has more code coverage than when I run coverage locally on master.

My branch is on the left, master is on the right. 
image

**Comment by pgrzesik** [comment]
I went through it once again and I believe it's just `coverage` playing with us here, it looks like it's all good, however, I think we should remove the last change with `{ image: true }` as it doesn't actually test anything extra as there are no new assertions in the test. Thanks for being patient with that PR :bow:

**Comment by ifitzsimmons** [comment]
> I went through it once again and I believe it's just `coverage` playing with us here, it looks like it's all good, however, I think we should remove the last change with `{ image: true }` as it doesn't actually test anything extra as there are no new assertions in the test. Thanks for being patient with that PR 🙇

I'm happy to remove it, but it does add coverage. I added another function object, but no additional calls to `fileExistsSyncStub` were made because of the following line
```javascript
if (functionObject.image) return;
```

If that object did not have an `image` property, we'd have to change the line in the test case:
```diff
- expect(fileExistsSyncStub.calledTwice).to.equal(true);
+ expect(fileExistsSyncStub.calledThrice).to.equal(true);
```

Let me know if you still want me to remove that line!! I'm happy to do so

**Comment by pgrzesik** [comment]
Hey @ifitzsimmons :wave: I understand that it might add coverage there, but the definition of that image is invalid (currently it expects a valid Docker Image URI) and there are no extra assertions added, so there's little value in added coverage if we don't really test the outcome of the tested path. I'll say let's remove the whole
```
second: {
          image: true,
        },
```
that was added, I don't think it will require changes in the assertions

**Comment by ifitzsimmons** [comment]
@pgrzesik Are you waiting on me for anything for this PR?

**Comment by pgrzesik** [comment]
Hello @ifitzsimmons, really sorry about it - I didn't realise that you pushed new changes because I'm monitoring review requests and comments via notifications and didn't get anything :( The best bet to ensure I or someone else sees the changes pushed is to request for re-review after completing changes.

Unfortunately, I see that there's one extra conflicting file that will need to be resolved before we can merge this one

**Comment by ifitzsimmons** [comment]
@pgrzesik That codecov issue resurfaced, can we ignore it again?

**Review by pgrzesik — APPROVED** [review]
Awesome job :raised_hands: Coverage looks good, your diff has 100% of coverage :100:

**Pull request merged**